File: python/camera_ui_common/common_utils/reactive.py
"""Reactive property utilities."""

from __future__ import annotations

from copy import deepcopy
from typing import Generic, TypeVar, cast

from camera_ui_sdk import BehaviorSubject, Observable, ReplaySubject, distinct_until_changed, share

T = TypeVar("T")


class ReactiveProperty(Generic[T]):
    """A reactive property that wraps a BehaviorSubject with observable access."""

    __subject: BehaviorSubject[T]
    observable: Observable[T]

    def __init__(
        self, initial_value: T | BehaviorSubject[T], observable: Observable[T] | None = None
    ) -> None:
        if isinstance(initial_value, BehaviorSubject):
            self.__subject = cast(BehaviorSubject[T], initial_value)
        else:
            self.__subject = BehaviorSubject(initial_value)

        self.observable = observable or self.__create_state_observable(self.__subject)

    @property
    def value(self) -> T:
        val = self.__subject.value

        if isinstance(val, dict | list) or hasattr(val, "__dict__"):
            return deepcopy(val)  # pyright: ignore[reportUnknownVariableType, reportUnknownArgumentType]

        return val

    def next(self, value: T) -> None:
        self.__subject.next(value)

    def complete(self) -> None:
        self.__subject.complete()

    def __create_state_observable(self, state_subject: BehaviorSubject[T]) -> Observable[T]:
        return state_subject.pipe(distinct_until_changed(), share(lambda: ReplaySubject(1)))

File: node/src/utils/reactive.ts
import { BehaviorSubject, ReplaySubject, distinctUntilChanged, share } from '@camera.ui/sdk';

import type { Observable } from '@camera.ui/sdk';

export class ReactiveProperty<T> {
  private _subject: BehaviorSubject<T>;

  public observable: Observable<T>;

  constructor(initialValue: T | BehaviorSubject<T>, oversable?: Observable<T>) {
    if (initialValue instanceof BehaviorSubject) {
      this._subject = initialValue;
    } else {
      this._subject = new BehaviorSubject(initialValue);
    }

    this.observable = oversable ?? this.#createStateObservable(this._subject);
  }

  get value(): T {
    const val = this._subject.getValue();

    if (typeof val === 'object') {
      return structuredClone(val);
    }

    return val;
  }

  public next(value: T): void {
    this._subject.next(value);
  }

  public complete(): void {
    this._subject.complete();
  }

  #createStateObservable<T>(stateSubject: BehaviorSubject<T>): Observable<T> {
    return stateSubject.pipe(distinctUntilChanged(), share({ connector: () => new ReplaySubject(1) }));
  }
}
